### PR TITLE
Adding rule for azurerm resource missing tags

### LINKED ISSUE #133: Required tags rule
A feature request:

Something similar to what was stated in terraform-linters/tflint#150 and implemented in terraform-linters/tflint#617 but for Azure resources.

In the config, you should be able to specify a tags attribute in the config block:

```
config {
    tags = ["foo", "bar"]
}
```

All the Azure resources with tags supported would be checked for tags matching these keys and a lint fail would occur for resources without those tags.

### PR DESCRIPTION
**Pull request by guybartal**
Resolves #133 based on the work done in [https://github.com/terraform-linters/tflint/pull/617](https://github.com/terraform-linters/tflint/pull/617) for AWS.

this PR allows you to get notified for resources with missing required tags:
```
rule "azurerm_resource_missing_tags" {
  enabled = true
  exclude = [] (list of resource types to exclude, optional)
  tags = ["Environment", "Department"]
}
```

### CODE REVIEW

**Comment by guybartal** [comment]
@wata727 , @hattan  can you please review?

File: README.md [hunk 1]
```diff
@@ -17,7 +17,7 @@ You can install the plugin by adding a config to `.tflint.hcl` and running `tfli
 ```hcl
 plugin "azurerm" {
     enabled = true
-    version = "0.17.1"
+    version = "0.17.2"
```
Comment (tamirkamara) [inline README.md:20]: Not sure what is the policy here, but if there's a new rule perhaps 0.18 is better?

Comment (guybartal) [reply README.md:20]: @wata727, moreover should I bump the version as part of the PR at all or is there another process for that?
I've noticed also that there's a changelog..

Comment (wata727) [reply README.md:20]: Version and changelog changes are usually done separately before release. No changes are required here.

File: integration/basic/.tflint.hcl [hunk 2]
```diff
@@ -1,3 +1,3 @@
 plugin "azurerm" {
   enabled = true
-}
+}
```
[the remark this answers is not in the document]
Comment (guybartal) [reply integration/basic/.tflint.hcl]: line was added

File: rules/azurerm_resource_missing_tags.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+// Link returns the rule reference link
+func (r *AzurermResourceMissingTagsRule) Link() string {
+	return project.ReferenceLink(r.Name())
+}
+
+// Check checks resources for missing tags
+func (r *AzurermResourceMissingTagsRule) Check(runner tflint.Runner) error {
+	// TODO: Write the implementation here. See this documentation for what tflint.Runner can do.
+	//       https://pkg.go.dev/github.com/terraform-linters/tflint-plugin-sdk/tflint#Runner
```
[the remark this answers is not in the document]
Comment (guybartal) [reply rules/azurerm_resource_missing_tags.go]: TODO comment removed

File: rules/azurerm_resource_missing_tags.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+type azurermResourceTagsRuleConfig struct {
+	Tags    []string `hclext:"tags"`
+	Exclude []string `hclext:"exclude,optional"`
+}
+
+const (
+	tagsAttributeName = "tags"
+	tagBlockName      = "tag"
```
[the remark this answers is not in the document]
Comment (guybartal) [reply rules/azurerm_resource_missing_tags.go]: ```tagBlockName``` removed

File: rules/azurerm_resource_missing_tags.go [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+	for _, tag := range config.Tags {
+		if _, ok := tags[tag]; !ok {
+			missing = append(missing, fmt.Sprintf("\"%s\"", tag))
+		}
+	}
+	if len(missing) > 0 {
+		sort.Strings(missing)
+		wanted := strings.Join(missing, ", ")
+		issue := fmt.Sprintf("The resource is missing the following tags: %s.", wanted)
+		runner.EmitIssue(r, issue, location)
```
[the remark this answers is not in the document]
Comment (guybartal) [reply rules/azurerm_resource_missing_tags.go]: error handled both in this function and in the caller using ```runner.EnsureNoError```

File: rules/tags/tags.go [hunk 6]
```diff
@@ -0,0 +1 @@
+package tags
```
[the remark this answers is not in the document]
Comment (guybartal) [reply rules/tags/tags.go:1]: ```go generate``` comment was added,
by the way I've noticed that golint is deprecated and should be replaced ([https://github.com/golang/go/issues/38968](https://github.com/golang/go/issues/38968))

File: project/main.go [hunk 7]
```diff
@@ -3,7 +3,7 @@ package project
 import "fmt"
 
 // Version is ruleset version
-const Version string = "0.17.1"
+const Version string = "0.17.2"
```
[the remark this answers is not in the document]
Comment (guybartal) [reply project/main.go]: reverted

File: rules/azurerm_resource_missing_tags.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (wata727) [reply rules/azurerm_resource_missing_tags.go]: The `EnsureNoError` is a helper function for handling warnings returned by `EvaluateExpr`. So, there is no need to use it here. Return an error and check it in the caller as follows:

```diff
diff --git a/rules/azurerm_resource_missing_tags.go b/rules/azurerm_resource_missing_tags.go
index c6fc048..07bcbc7 100644
--- a/rules/azurerm_resource_missing_tags.go
+++ b/rules/azurerm_resource_missing_tags.go
@@ -87,15 +87,16 @@ func (r *AzurermResourceMissingTagsRule) Check(runner tflint.Runner) error {
 				wantType := cty.Map(cty.String)
 				err := runner.EvaluateExpr(attribute.Expr, &resourceTags, &tflint.EvaluateExprOption{WantType: &wantType})
 				err = runner.EnsureNoError(err, func() error {
-					r.emitIssue(runner, resourceTags, config, attribute.Expr.Range())
-					return nil
+					return r.emitIssue(runner, resourceTags, config, attribute.Expr.Range())
 				})
 				if err != nil {
 					return err
 				}
 			} else {
 				logger.Debug("checking", "resource type", resource.Labels[0], "resource name", resource.Labels[1])
-				r.emitIssue(runner, map[string]string{}, config, resource.DefRange)
+				if err := r.emitIssue(runner, map[string]string{}, config, resource.DefRange); err != nil {
+					return err
+				}
 			}
 		}
 	}
@@ -103,7 +104,7 @@ func (r *AzurermResourceMissingTagsRule) Check(runner tflint.Runner) error {
 	return nil
 }
 
-func (r *AzurermResourceMissingTagsRule) emitIssue(runner tflint.Runner, tags map[string]string, config azurermResourceTagsRuleConfig, location hcl.Range) {
+func (r *AzurermResourceMissingTagsRule) emitIssue(runner tflint.Runner, tags map[string]string, config azurermResourceTagsRuleConfig, location hcl.Range) error {
 	var missing []string
 	for _, tag := range config.Tags {
 		if _, ok := tags[tag]; !ok {
@@ -114,8 +115,9 @@ func (r *AzurermResourceMissingTagsRule) emitIssue(runner tflint.Runner, tags ma
 		sort.Strings(missing)
 		wanted := strings.Join(missing, ", ")
 		issue := fmt.Sprintf("The resource is missing the following tags: %s.", wanted)
-		runner.EmitIssue(r, issue, location)
+		return runner.EmitIssue(r, issue, location)
 	}
+	return nil
 }
 
 func stringInSlice(a string, list []string) bool {

```

File: rules/tags/tags.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (wata727) [reply rules/tags/tags.go:1]: Yeah, we'll need to replace it with [golangci-lint](https://github.com/golangci/golangci-lint) or simply turn it off. For rulesets, godoc is not so important, so we may not need to enforce this rule.

**Review by wata727 — COMMENTED** [review]
Good, please check the error handling lastly.

File: rules/azurerm_resource_missing_tags.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (guybartal) [reply rules/azurerm_resource_missing_tags.go]: got it, thanks!
fixed.

**Comment by guybartal** [comment]
> Good, please check the error handling lastly.

did I fix it correctly?

**Review by wata727 — APPROVED** [review]
Perfect. Thanks!

**Comment by guybartal** [comment]
> Perfect. Thanks!

Can you please create a new release so we could use this rule?

**Comment by wata727** [comment]
Sure. I'm planning to release a new version within a few days.

**Pull request merged**